File: panel-go/api/tunnels.go
package api

import (
	"encoding/json"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"weavenet/panel/cache"
	"weavenet/panel/database"
	"weavenet/panel/middleware"
	"weavenet/panel/models"
	"weavenet/panel/services"
	"weavenet/panel/utils"
)

type tunnelPayload struct {
	Name          string         `json:"name"`
	NodeID        uint           `json:"node_id"`
	Type          string         `json:"type"`
	LocalIP       string         `json:"local_ip"`
	LocalPort     int            `json:"local_port"`
	RemotePort    *int           `json:"remote_port"`
	Subdomain     *string        `json:"subdomain"`
	CustomDomain  *string        `json:"custom_domain"`
	KCP           bool           `json:"kcp"`
	Encryption    bool           `json:"encryption"`
	Compression   bool           `json:"compression"`
	SecretKey     *string        `json:"secret_key"`
	LoadBalancers []map[string]any `json:"load_balancers"`
}

// serializeTunnel 序列化隧道（含实时状态与公网地址）。
func serializeTunnel(t *models.Tunnel, node *models.Node) gin.H {
	var nodeInfo gin.H
	if node != nil && node.ID > 0 {
		nodeInfo = gin.H{
			"id": node.ID, "name": node.Name, "address": node.Address, "port": node.Port,
			"status": node.Status, "speed_limit_mbps": node.SpeedLimitMbps,
			"last_heartbeat_at": utils.TimeFmt(node.LastHeartbeatAt),
		}
	} else {
		nodeInfo = nil
	}
	runtime, _ := cache.C.Get(cache.RuntimeKey(t.ID))
	online := false
	connections := 0
	if rt, ok := runtime.(cache.TunnelRuntime); ok {
		online = rt.Online
		connections = rt.Connections
	}
	today := utils.Today()
	tr, _ := cache.GetTraffic(today, t.ID)
	addr := ""
	if node != nil {
		addr = services.BuildPublicAddress(t, node)
	}
	return gin.H{
		"id": t.ID, "user_id": t.UserID, "node_id": t.NodeID, "name": t.Name, "type": t.Type,
		"local_ip": t.LocalIP, "local_port": t.LocalPort, "remote_port": t.RemotePort,
		"subdomain": t.Subdomain, "custom_domain": t.CustomDomain, "kcp": t.KCP,
		"encryption": t.Encryption, "compression": t.Compression, "secret_key": t.SecretKey,
		"load_balancers": services.ParseLB(t.LoadBalancers), "status": t.Status,
		"status_detail": t.StatusDetail, "created_at": utils.TimeFmtV(t.CreatedAt),
		"node": nodeInfo, "online": online, "connections": connections,
		"today_in": tr.In, "today_out": tr.Out, "public_address": addr,
	}
}

// ListTunnels 隧道列表。
func ListTunnels(c *gin.Context) {
	user := middleware.GetUser(c)
	db := database.DB
	q := db.Where("user_id = ?", user.ID)
	if t := c.Query("type"); t != "" {
		q = q.Where("type = ?", t)
	}
	var tunnels []models.Tunnel
	q.Order("id desc").Find(&tunnels)
	items := make([]gin.H, 0, len(tunnels))
	for _, t := range tunnels {
		var node models.Node
		db.First(&node, t.NodeID)
		items = append(items, serializeTunnel(&t, &node))
	}
	utils.OK(c, items, "ok")
}

// getOwnedTunnel 获取用户所属隧道。
func getOwnedTunnel(c *gin.Context) (*models.Tunnel, *models.Node, *utils.BizError) {
	user := middleware.GetUser(c)
	id, err := strconv.Atoi(c.Param("tunnel_id"))
	if err != nil {
		return nil, nil, utils.NotFound("隧道不存在")
	}
	db := database.DB
	var t models.Tunnel
	if err := db.Where("id = ? AND user_id = ?", id, user.ID).First(&t).Error; err != nil {
		return nil, nil, utils.NotFound("隧道不存在")
	}
	var node models.Node
	db.First(&node, t.NodeID)
	return &t, &node, nil
}

// GetTunnel 隧道详情。
func GetTunnel(c *gin.Context) {
	t, node, be := getOwnedTunnel(c)
	if be != nil {
		utils.Fail(c, be)
		return
	}
	utils.OK(c, serializeTunnel(t, node), "ok")
}

// CreateTunnel 创建隧道。
func CreateTunnel(c *gin.Context) {
	var p tunnelPayload
	if err := c.ShouldBindJSON(&p); err != nil {
		utils.Fail(c, utils.ErrBiz(422, 0, "参数错误"))
		return
	}
	db := database.DB
	user := middleware.GetUser(c)
	if p.Name == "" || p.NodeID == 0 || p.LocalPort < 1 {
		utils.Fail(c, utils.ErrBiz(400, 0, "请填写名称、节点与本地端口"))
		return
	}
	var node models.Node
	if err := db.First(&node, p.NodeID).Error; err != nil {
		utils.Fail(c, utils.NotFound("节点不存在"))
		return
	}
	if node.Status != "online" {
		utils.Fail(c, utils.ErrBiz(400, 3004, "节点离线或维护中，无法创建隧道"))
		return
	}
	if err := services.CheckTunnelLimit(db, user, &user.Plan); err != nil {
		utils.Fail(c, utils.AsBiz(err))
		return
	}
	remotePort := p.RemotePort
	if p.Type == "tcp" || p.Type == "udp" || p.Type == "kcp" || p.Type == "loadbalance" {
		if remotePort == nil {
			port, err := services.AllocateRemotePort(db, p.NodeID)
			if err != nil {
				utils.Fail(c, utils.AsBiz(err))
				return
			}
			remotePort = &port
		} else if !services.CheckRemotePortAvailable(db, p.NodeID, *remotePort, 0) {
			utils.Fail(c, utils.ErrBiz(409, 3001, "远程端口已被占用"))
			return
		}
	}
	if p.Subdomain != nil && *p.Subdomain != "" {
		if !services.CheckSubdomainAvailable(db, *p.Subdomain, 0) {
			utils.Fail(c, utils.ErrBiz(409, 3002, "子域名已被占用"))
			return
		}
	}
	lbStr := ""
	if len(p.LoadBalancers) > 0 {
		b, _ := json.Marshal(p.LoadBalancers)
		lbStr = string(b)
	}
	t := models.Tunnel{
		UserID: user.ID, NodeID: p.NodeID, Name: p.Name, Type: p.Type,
		LocalIP: orDefault(p.LocalIP, "127.0.0.1"), LocalPort: p.LocalPort,
		RemotePort: remotePort, Subdomain: p.Subdomain, CustomDomain: p.CustomDomain,
		KCP: p.KCP, Encryption: p.Encryption, Compression: p.Compression,
		SecretKey: p.SecretKey, LoadBalancers: lbStr, Status: "stopped",
	}
	if t.Type == "stcp" || t.Type == "xtcp" {
		sk := utils.GenSecretKey()
		t.SecretKey = &sk
	}
	if err := db.Create(&t).Error; err != nil {
		utils.Fail(c, utils.AsBiz(err))
		return
	}
	utils.OK(c, serializeTunnel(&t, &node), "隧道创建成功")
}

// UpdateTunnel 更新隧道。
func UpdateTunnel(c *gin.Context) {
	t, _, be := getOwnedTunnel(c)
	if be != nil {
		utils.Fail(c, be)
		return
	}
	var p struct {
		Name         string         `json:"name"`
		LocalIP      string         `json:"local_ip"`
		LocalPort    int            `json:"local_port"`
		RemotePort   *int           `json:"remote_port"`
		Subdomain    *string        `json:"subdomain"`
		CustomDomain *string        `json:"custom_domain"`
		KCP          *bool          `json:"kcp"`
		Encryption   *bool          `json:"encryption"`
		Compression  *bool          `json:"compression"`
		SecretKey    *string        `json:"secret_key"`
		LoadBalancers []map[string]any `json:"load_balancers"`
	}
	if err := c.ShouldBindJSON(&p); err != nil {
		utils.Fail(c, utils.ErrBiz(422, 0, "参数错误"))
		return
	}
	db := database.DB
	if p.Name != "" {
		t.Name = p.Name
	}
	if p.LocalIP != "" {
		t.LocalIP = p.LocalIP
	}
	if p.LocalPort > 0 {
		t.LocalPort = p.LocalPort
	}
	if p.RemotePort != nil {
		if !services.CheckRemotePortAvailable(db, t.NodeID, *p.RemotePort, t.ID) {
			utils.Fail(c, utils.ErrBiz(409, 3001, "远程端口已被占用"))
			return
		}
		t.RemotePort = p.RemotePort
	}
	if p.Subdomain != nil && *p.Subdomain != "" {
		if !services.CheckSubdomainAvailable(db, *p.Subdomain, 0) {
			utils.Fail(c, utils.ErrBiz(409, 3002, "子域名已被占用"))
			return
		}
		t.Subdomain = p.Subdomain
	}
	if p.CustomDomain != nil {
		t.CustomDomain = p.CustomDomain
	}
	if p.KCP != nil {
		t.KCP = *p.KCP
	}
	if p.Encryption != nil {
		t.Encryption = *p.Encryption
	}
	if p.Compression != nil {
		t.Compression = *p.Compression
	}
	if p.SecretKey != nil {
		t.SecretKey = p.SecretKey
	}
	if len(p.LoadBalancers) > 0 {
		b, _ := json.Marshal(p.LoadBalancers)
		t.LoadBalancers = string(b)
	}
	db.Save(t)
	utils.OK(c, serializeTunnel(t, nil), "隧道更新成功")
}

// DeleteTunnel 删除隧道。
func DeleteTunnel(c *gin.Context) {
	t, _, be := getOwnedTunnel(c)
	if be != nil {
		utils.Fail(c, be)
		return
	}
	db := database.DB
	db.Model(&models.Domain{}).Where("tunnel_id = ?", t.ID).Update("tunnel_id", nil)
	db.Delete(t)
	cache.C.Del(cache.RuntimeKey(t.ID))
	cache.C.Del(cache.WantKey(t.ID))
	utils.NoContent(c)
}

// StartTunnel 启动隧道。
func StartTunnel(c *gin.Context) {
	t, _, be := getOwnedTunnel(c)
	if be != nil {
		utils.Fail(c, be)
		return
	}
	var node models.Node
	database.DB.First(&node, t.NodeID)
	if node.Status != "online" {
		utils.Fail(c, utils.ErrBiz(400, 3004, "节点离线或维护中"))
		return
	}
	t.Status = "running"
	database.DB.Save(t)
	// 写启停指令（frps 轮询感知）
	cmd, _ := json.Marshal(gin.H{"action": "start", "force": false})
	cache.C.Set(cache.WantKey(t.ID), cmd, time.Hour)
	utils.OKMsg(c, "隧道启动指令已下发")
}

// StopTunnel 停止隧道。
func StopTunnel(c *gin.Context) {
	t, _, be := getOwnedTunnel(c)
	if be != nil {
		utils.Fail(c, be)
		return
	}
	t.Status = "stopped"
	database.DB.Save(t)
	cmd, _ := json.Marshal(gin.H{"action": "stop", "force": false})
	cache.C.Set(cache.WantKey(t.ID), cmd, time.Hour)
	utils.OKMsg(c, "隧道停止指令已下发")
}

// GetTunnelConfig 生成 frpc.toml 配置。
func GetTunnelConfig(c *gin.Context) {
	t, node, be := getOwnedTunnel(c)
	if be != nil {
		utils.Fail(c, be)
		return
	}
	user := middleware.GetUser(c)
	cfg := services.GenerateFrpcConfig(database.DB, t, user, node)
	utils.OK(c, gin.H{"config": cfg}, "ok")
}

// GetTunnelStatus 隧道实时状态。
func GetTunnelStatus(c *gin.Context) {
	t, _, be := getOwnedTunnel(c)
	if be != nil {
		utils.Fail(c, be)
		return
	}
	rt, _ := cache.C.Get(cache.RuntimeKey(t.ID))
	online, connections := false, 0
	var in, out int64
	ts := ""
	if r, ok := rt.(cache.TunnelRuntime); ok {
		online, connections = r.Online, r.Connections
		in, out, ts = r.In, r.Out, r.Ts
	}
	tr, _ := cache.GetTraffic(utils.Today(), t.ID)
	utils.OK(c, gin.H{
		"status": t.Status, "online": online, "connections": connections,
		"in_bytes": in, "out_bytes": out, "today_in": tr.In, "today_out": tr.Out, "ts": ts,
	}, "ok")
}

func orDefault(s, def string) string {
	if s == "" {
		return def
	}
	return s
}

File: panel/app/api/tunnels.py
"""隧道 API：CRUD、启停、配置生成、状态。"""
from __future__ import annotations

import json

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from ..core.database import get_db
from ..core.deps import get_current_user
from ..core.errors import BizError, ConflictError, NotFoundError, success
from ..core.redis_client import redis_client
from ..models import Domain, Node, Tunnel, User
from ..schemas.schemas import TunnelCreate, TunnelUpdate
from ..services.frp_config import (
    allocate_remote_port,
    build_public_address,
    check_remote_port_available,
    check_subdomain_available,
    check_tunnel_limit,
    generate_frpc_config,
)

router = APIRouter(prefix="/api/tunnels", tags=["隧道"])


def _serialize_tunnel(tunnel: Tunnel) -> dict:
    """序列化隧道并附加实时状态。"""
    runtime = redis_client.get_tunnel_runtime(tunnel.id) or {}
    node = tunnel.node
    return {
        "id": tunnel.id,
        "user_id": tunnel.user_id,
        "node_id": tunnel.node_id,
        "name": tunnel.name,
        "type": tunnel.type,
        "local_ip": tunnel.local_ip,
        "local_port": tunnel.local_port,
        "remote_port": tunnel.remote_port,
        "subdomain": tunnel.subdomain,
        "custom_domain": tunnel.custom_domain,
        "kcp": tunnel.kcp,
        "encryption": tunnel.encryption,
        "compression": tunnel.compression,
        "secret_key": tunnel.secret_key,
        "load_balancers": tunnel.load_balancers,
        "status": tunnel.status,
        "status_detail": tunnel.status_detail,
        "created_at": tunnel.created_at.isoformat() if tunnel.created_at else None,
        "node": {
            "id": node.id,
            "name": node.name,
            "address": node.address,
            "port": node.port,
            "status": node.status,
        }
        if node
        else None,
        "online": bool(runtime.get("online", False)) or tunnel.status == "running",
        "connections": int(runtime.get("connections", 0)),
        "today_in": int(runtime.get("in", 0)),
        "today_out": int(runtime.get("out", 0)),
        "public_address": build_public_address(tunnel, node),
    }


@router.get("")
def list_tunnels(
    type: str | None = None,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """隧道列表。"""
    q = db.query(Tunnel).filter(Tunnel.user_id == user.id)
    if type:
        q = q.filter(Tunnel.type == type)
    tunnels = q.order_by(Tunnel.id.desc()).all()
    return success([_serialize_tunnel(t) for t in tunnels])


@router.post("")
def create_tunnel(
    payload: TunnelCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """创建隧道。"""
    db.refresh(user)
    plan = user.plan
    if plan is None:
        raise BizError(400, 2001, "套餐配置异常，请联系管理员")
    check_tunnel_limit(user, plan)

    node = db.get(Node, payload.node_id)
    if node is None:
        raise NotFoundError(message="节点不存在")
    if node.status != "online":
        raise BizError(400, 3004, "节点离线或维护中，请选择其他节点")

    # 远程端口：自动分配或校验指定端口
    remote_port = payload.remote_port
    if payload.type in ("tcp", "udp", "stcp", "xtcp", "kcp", "loadbalance"):
        if remote_port is None:
            remote_port = allocate_remote_port(db, node.id)
        elif not check_remote_port_available(db, node.id, remote_port):
            raise ConflictError(3001, "远程端口已被占用，请更换端口")
    else:
        # http/https 使用域名路由，无需远程端口（可自动分配占位）
        remote_port = None

    # 域名相关校验
    if payload.subdomain:
        if not check_subdomain_available(db, payload.subdomain):
            raise ConflictError(3002, "子域名已被占用，请更换")
    if payload.custom_domain:
        existed = (
            db.query(Tunnel)
            .filter(Tunnel.custom_domain == payload.custom_domain, Tunnel.user_id != user.id)
            .first()
        )
        if existed:
            raise ConflictError(3002, "自定义域名已被其他用户使用")

    tunnel = Tunnel(
        user_id=user.id,
        node_id=node.id,
        name=payload.name,
        type=payload.type,
        local_ip=payload.local_ip,
        local_port=payload.local_port,
        remote_port=remote_port,
        subdomain=payload.subdomain,
        custom_domain=payload.custom_domain,
        kcp=payload.kcp,
        encryption=payload.encryption,
        compression=payload.compression,
        secret_key=payload.secret_key,
        load_balancers=json.dumps(payload.load_balancers, ensure_ascii=False)
        if payload.load_balancers
        else None,
        status="stopped",
    )
    db.add(tunnel)
    db.commit()
    db.refresh(tunnel)
    return success(_serialize_tunnel(tunnel), "隧道创建成功", 201)


@router.get("/{tunnel_id}")
def tunnel_detail(
    tunnel_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """隧道详情。"""
    tunnel = _get_owned_tunnel(db, user, tunnel_id)
    return success(_serialize_tunnel(tunnel))


@router.put("/{tunnel_id}")
def update_tunnel(
    tunnel_id: int,
    payload: TunnelUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """修改隧道。"""
    tunnel = _get_owned_tunnel(db, user, tunnel_id)

    if payload.remote_port is not None and payload.remote_port != tunnel.remote_port:
        if not check_remote_port_available(db, tunnel.node_id, payload.remote_port, tunnel.id):
            raise ConflictError(3001, "远程端口已被占用，请更换端口")
        tunnel.remote_port = payload.remote_port

    if payload.subdomain is not None and payload.subdomain != tunnel.subdomain:
        if not check_subdomain_available(db, payload.subdomain):
            raise ConflictError(3002, "子域名已被占用，请更换")
        tunnel.subdomain = payload.subdomain or None

    for field in ("name", "local_ip", "local_port", "custom_domain", "kcp", "encryption", "compression", "secret_key"):
        value = getattr(payload, field, None)
        if value is not None:
            setattr(tunnel, field, value)

    if payload.load_balancers is not None:
        tunnel.load_balancers = json.dumps(payload.load_balancers, ensure_ascii=False)

    db.commit()
    db.refresh(tunnel)
    return success(_serialize_tunnel(tunnel), "隧道已更新")


@router.delete("/{tunnel_id}", status_code=204)
def delete_tunnel(
    tunnel_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """删除隧道。"""
    tunnel = _get_owned_tunnel(db, user, tunnel_id)
    db.query(Domain).filter(Domain.tunnel_id == tunnel.id).update({"tunnel_id": None})
    redis_client.delete(f"tunnel:runtime:{tunnel.id}")
    db.delete(tunnel)
    db.commit()


@router.post("/{tunnel_id}/start")
def start_tunnel(
    tunnel_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """启动隧道。"""
    tunnel = _get_owned_tunnel(db, user, tunnel_id)
    node = tunnel.node
    if node is None or node.status == "offline":
        raise BizError(400, 3004, "节点离线，无法启动隧道")
    if node.status == "maintenance":
        raise BizError(400, 3004, "节点维护中，请稍后再试")
    tunnel.status = "running"
    tunnel.status_detail = "已请求启动"
    db.commit()
    # 通知 frps 热更新（由 agent 每 10s 拉取，此处刷新 Redis 缓存加速感知）
    redis_client.set_json(f"tunnel:want:{tunnel.id}", 3600, {"action": "start"})
    return success(_serialize_tunnel(tunnel), "启动指令已下发")


@router.post("/{tunnel_id}/stop")
def stop_tunnel(
    tunnel_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """停止隧道。"""
    tunnel = _get_owned_tunnel(db, user, tunnel_id)
    tunnel.status = "stopped"
    tunnel.status_detail = "已请求停止"
    db.commit()
    redis_client.set_json(f"tunnel:want:{tunnel.id}", 3600, {"action": "stop"})
    return success(_serialize_tunnel(tunnel), "停止指令已下发")


@router.post("/{tunnel_id}/config")
def tunnel_config(
    tunnel_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """生成 frpc 配置。"""
    tunnel = _get_owned_tunnel(db, user, tunnel_id)
    node = tunnel.node
    if node is None:
        raise NotFoundError(message="节点不存在")
    config = generate_frpc_config(tunnel, user, node)
    return success({"config": config, "filename": f"frpc-{tunnel.name}.toml"})


@router.get("/{tunnel_id}/status")
def tunnel_status(
    tunnel_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """查询隧道状态。"""
    tunnel = _get_owned_tunnel(db, user, tunnel_id)
    runtime = redis_client.get_tunnel_runtime(tunnel.id) or {}
    return success(
        {
            "tunnel_id": tunnel.id,
            "status": tunnel.status,
            "status_detail": tunnel.status_detail,
            "online": bool(runtime.get("online", False)),
            "connections": int(runtime.get("connections", 0)),
            "today_in": int(runtime.get("in", 0)),
            "today_out": int(runtime.get("out", 0)),
            "updated_at": runtime.get("ts"),
        }
    )


def _get_owned_tunnel(db: Session, user: User, tunnel_id: int) -> Tunnel:
    """获取当前用户所属隧道，ORM 层强制 user_id 过滤防越权。"""
    tunnel = (
        db.query(Tunnel)
        .filter(Tunnel.id == tunnel_id, Tunnel.user_id == user.id)
        .first()
    )
    if tunnel is None:
        raise NotFoundError(message="隧道不存在")
    return tunnel
